source/extensions/common/wasm/wasm.cc

```diff
@@ -535,7 +535,48 @@ Word ___syscall54Handler(void*, Word, Word) { throw WasmException("emscripten sy
 
 Word ___syscall140Handler(void*, Word, Word) { throw WasmException("emscripten syscall140"); }
 
-Word ___syscall146Handler(void*, Word, Word) { throw WasmException("emscripten syscall146"); }
+// Implementation of writev() syscall that redirects stdout/stderr to Envoy logs.
+// ssize_t writev(int fd, const struct iovec *iov, int iovcnt);
+Word ___syscall146Handler(void* raw_context, Word, Word syscall_args_ptr) {
+  auto context = WASM_CONTEXT(raw_context);
+
+  // Read syscall args.
+  auto memslice = context->wasmVm()->getMemory(syscall_args_ptr, 3 * sizeof(uint32_t));
+  const uint32_t* syscall_args = reinterpret_cast<const uint32_t*>(memslice.data());
+
+  spdlog::level::level_enum log_level;
+  switch (syscall_args[0] /* fd */) {
+  case 1 /* stdout */:
+    log_level = spdlog::level::info;
+    break;
+  case 2 /* stderr */:
+    log_level = spdlog::level::err;
+    break;
+  default:
+    throw WasmException("emscripten syscall146 (writev)");
+  }
+
+  std::string s;
+  for (size_t i = 0; i < syscall_args[2] /* iovcnt */; i++) {
+    memslice = context->wasmVm()->getMemory(syscall_args[1] /* iov */ + i * 2 * sizeof(uint32_t),
+                                            2 * sizeof(uint32_t));
+    const uint32_t* iovec = reinterpret_cast<const uint32_t*>(memslice.data());
+    if (iovec[1] /* size */) {
+      memslice = context->wasmVm()->getMemory(iovec[0] /* data */, iovec[1] /* size */);
+      s.append(memslice.data(), memslice.size());
+    }
+  }
+
+  size_t written = s.size();
+  if (written) {
+    // Remove trailing newline from the logs, if any.
```

Comment:
I would be surprised when I get the same log despite I write with "\n" or without "\n". TBH I don't think it is the syscall layer to handle this logic.

Reply:
We're intercepting writes to stdout/stderr and writing them to Envoy logs, where each entry already has a newline at the end. If we don't remove those trailing newlines, then each intercepted write to stdout/stderr will be followed by an empty line.

I don't think that removing empty lines is more intrusive than prefixing those writes with a timestamp, filename and `wasm log:`, so I'm going to leave this as-is.